crates/bin/pd/src/migrate/testnet72.rs

```diff
@@ -0,0 +1,206 @@
+                    height: epoch_starting_height,
+                }
+                .into_request(),
+            )
+            .await?
+            .into_inner()
+            .epoch
+            .expect(&format!(
+                "epoch at height {} should be present",
+                epoch_starting_height
+            ))
+            .index)
+    }
+
+    /// Translate the protobuf for a BSOD by populating the correct data and emptying the
+    /// deprecated field.
+    #[allow(deprecated)]
+    async fn translate_bsod(
+        &self,
+        bsod: pb::dex::v1::BatchSwapOutputData,
+    ) -> anyhow::Result<pb::dex::v1::BatchSwapOutputData> {
+        let sct_position_prefix: u64 = {
+            let epoch = self
+                .epoch_height_to_index(bsod.epoch_starting_height)
+                .await?;
+            Position::from((
+                u16::try_from(epoch).expect("epoch should fit in 16 bits"),
+                u16::try_from(bsod.height - bsod.epoch_starting_height)
+                    .expect("block index should fit in 16 bits"),
+                0,
+            ))
+            .into()
+        };
+        Ok(pb::dex::v1::BatchSwapOutputData {
+            sct_position_prefix,
+            epoch_starting_height: Default::default(),
+            ..bsod
+        })
+    }
+
+    async fn translate_compact_block(
+        &self,
+        compact_block: pb::compact_block::v1::CompactBlock,
+    ) -> anyhow::Result<pb::compact_block::v1::CompactBlock> {
+        let mut swap_outputs = Vec::with_capacity(compact_block.swap_outputs.len());
+        for bsod in compact_block.swap_outputs {
+            swap_outputs.push(self.translate_bsod(bsod).await?);
+        }
+        Ok(pb::compact_block::v1::CompactBlock {
+            swap_outputs,
+            ..compact_block
+        })
+    }
+}
+
+/// Translate all of the BSODs inside dex storage to the new format.
+async fn translate_dex_storage(
+    ctx: Context,
+    delta: &mut StateDelta<Snapshot>,
+) -> anyhow::Result<()> {
+    let mut stream = delta.prefix_raw("dex/output/");
+    while let Some(r) = stream.next().await {
+        let (key, bsod_bytes) = r?;
+        let bsod = pb::dex::v1::BatchSwapOutputData::decode(bsod_bytes.as_slice())?;
+        let bsod = ctx.translate_bsod(bsod).await?;
+        delta.put_proto(key, bsod);
+    }
+    Ok(())
+}
+
+/// Translate all of the compact block storage to hold the new BSOD data inside the compact blocks.
+async fn translate_compact_block_storage(
+    ctx: Context,
+    delta: &mut StateDelta<Snapshot>,
+) -> anyhow::Result<()> {
+    let mut stream = delta.prefix_raw("compactblock/");
+    while let Some(r) = stream.next().await {
+        let (key, compactblack_bytes): (String, Vec<u8>) = r?;
+        let block = pb::compact_block::v1::CompactBlock::decode(compactblack_bytes.as_slice())?;
+        let block = ctx.translate_compact_block(block).await?;
+        delta.put_proto(key, block);
```

Comment:
Here the structure is correct, but we need to be reading from/writing to nonverifiable storage (the key looks correct).